### ISSUE TITLE
Implement socket communication via protocol buffers

### ISSUE THREAD

**Issue by douglasrizzo** [santiontanon/microrts#49]
Hello @santiontanon . You have mentioned to me in another medium that, when communicating to and from MicroRTS using sockets, building and passing XML and JSON messages creates a bottleneck in simulation time.

I have seen a tendency in the community to opt for [protocol buffers](https://developers.google.com/protocol-buffers/) (protobuf) when communication time is a priority over human-readability.

Basically, you create a `.proto` text file which dictates the message format, compile it using a tool that generates classes in multiple languages, and use these classes to create interchangeable objects, which can be serialized in byte arrays, for example.

I have found [this comparison](https://auth0.com/blog/beating-json-performance-with-protobuf/) between JSON and Protobuf which indicates that the later is up to 5x times faster.

I was working through the tutorials and ended up creating a `.proto` file for a MicroRTS game state and was thinking about testing it sometime soon.

If successful, would there be interest to include something like this into MicroRTS?

```proto
syntax = "proto3";

option java_package = "ai.socket.protobuf";
option java_outer_classname = "GameStateProtobuf";

message unit {
  string type = 1;
  int32 ID = 2;
  int32 player = 3;
  int32 x = 4;
  int32 y = 5;
  int32 resources = 6;
  int32 hitpoints = 7;
}

message player {
  int32 ID = 1;
  int32 resources = 2;
}

message PGS {
  int32 width = 1;
  int32 height = 2;
  string terrain = 3;
  repeated player players = 4;
  repeated unit units = 5;
}

message action {
  int32 type = 1;
  int32 parameter = 2;
}

message actions {
  int32 ID = 1;
  int32 time = 2;
  action action = 3;
}

message state {
  int32 time = 1;
  PGS pgs = 2;
  repeated actions actions = 3;
}
```

**Comment by santiontanon** [comment]
Hi Douglas, yes! I am very familiar with protobuffs, as I use them regularly in my everyday work at Google. That would be great! As I mentioned on the Gradle-related pull request, if we go with this, I'd like to do it in a way that we do not depend on an external package, so I need to think about how to do this. But yes, it's something I'd definitively be interested in!

**Comment by douglasrizzo** [comment]
I haven't used protobufs before but I believe that, once the classes derived from the protos are generated, you actually don't need any dependencies. All the methods necessary to serialize and deserialize objects are contained inside the classes. It would be good to add the protos to the repo, though.

**Comment by santiontanon** [comment]
Yes, I agree. That would be a good solution. Add the protobuffs to the repo (and maybe a build script for whoever wants to rebuild), but include the result of compiling them in to a Java class!

**Comment by douglasrizzo** [comment]
OK, I'll investigate this further in a fork.

**Comment by santiontanon** [comment]
Awesome, thanks a lot! (apologies for rejecting your other pull request concerning Gradle, but that was something I really didn't want to do).

**Comment by douglasrizzo** [comment]
That pull request was not mine!

**Comment by santiontanon** [comment]
oh, true! got confused, since you replied there, my bad! :)
